src/ctapipe/image/modifications.py

```diff
@@ -100,6 +106,179 @@ def _smear_psf_randomly(
+    simulation settings) as the file to which the noise is to be added,
+    but containing only NSB noise (electronic noise should be switched off)
+
+    The number of simulated noise events per telescope in the NSB file must
+    be at least twice the number of waveforms ("nsb_level") from that file
+    that we want to add up. If the NSB file is produced with a level of 25%
+    of dark NSB, and we want to simulate 10x dark NSB, then nsb_level=40 (
+    =10/0.25) and the file must contain at least 80 events. This is to
+    guarantee that the different noise waveforms are not too correlated
+
+    """
+
+    nsb_file = Path(
+        default_value=None,
+        help="Path to a dedicated NSB-only sim_telarray file",
+    ).tag(config=True)
+
+    nsb_level = Int(
+        default_value=1,
+        help=(
+            "Number of random instances of the NSB waveforms from the "
+            "NSB file to be added up to the waveform"
+        ),
+    ).tag(config=True)
+
+    n_noise_realizations = Int(
+        default_value=100,
+        help=(
+            "Number of different realizations of the total NSB waveform to "
+            "be created per pixel"
+        ),
+    ).tag(config=True)
+
+    rng_seed = Int(default_value=1, help="Seed for the random number generator").tag(
+        config=True
+    )
+
+    total_noise = dict()
+    # One key per tel_id, each of them is an array of shape
+    # [n_noise_realizations, ngains, npixels, nsamples]
+
+    def __init__(
+        self, subarray: SubarrayDescription, config=None, parent=None, **kwargs
+    ):
+        """
+
+        Parameters
+        ----------
+        subarray: SubarrayDescription
+            Description of the subarray. Provides information about the
+            camera which are useful in calibration. Also required for
+            configuring the TelescopeParameter traitlets.
+        config: traitlets.loader.Config
+            Configuration specified by config file or cmdline arguments.
+            Used to set traitlet values.
+            This is mutually exclusive with passing a ``parent``.
+        parent: ctapipe.core.Component or ctapipe.core.Tool
+            Parent of this component in the configuration hierarchy,
+            this is mutually exclusive with passing ``config``
+
+        """
+
+        super().__init__(subarray=subarray, config=config, parent=parent, **kwargs)
+        self.rng = np.random.default_rng(self.rng_seed)
+
+        # Read in the waveforms in the NSB-only file. Store in a dictionary
+        # with one key per telescope, containing an array [n_events, n_gains,
+        # n_pixels, n_samples]
+        source = EventSource(input_url=self.nsb_file, skip_calibration_events=False)
+        nsb_database = dict()  # [nevents, ngains, npixels, nsamples]
+        for event in source:
+            for tel_id in event.trigger.tels_with_trigger:
+                if tel_id in nsb_database:
```

Comment:
much cheaper and simpler to just accumulate into a list and then stack once in the end.

Also doesn't need the if.